### PR TITLE
Add Command to Set Logging/Filter Rules at Runtime

### PR DESCRIPTION
**Pull request by Nerixyz**
# Description

This PR adds the `/c2:loggingrules` command. You invoke it like this: `/c2:loggingrules chatterino.*.debug=true chatterino.http.debug=false`. This will apply both rules. I'm not too sure about the command name/prefix/namespace (my thinking was to have this behind a namespace/prefix, so it's clear that this is a Chatterino command).

This command is mainly for debugging release builds without restarting - so that you get more logs when you attach a debugger/console or even a dedicated program capturing logs.

### CODE REVIEW

File: src/controllers/commands/builtin/chatterino/Debugging.hpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+struct CommandContext;
+
+namespace commands {
+
+    QString setLoggingRules(const CommandContext &ctx);
+
+}  // namespace commands
+
+}  // namespace chatterino
```
Comment (pajlada) [inline src/controllers/commands/builtin/chatterino/Debugging.hpp:15]: I would rather the forward-declaration is done in a separate namespace definition - see ShieldMode.hpp

File: src/controllers/commands/builtin/chatterino/Debugging.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            "the format on the rules, see "
+            "https://doc.qt.io/qt-6/"
+            "qloggingcategory.html#configuring-categories"));
+        return {};
+    }
+
+    QLoggingCategory::setFilterRules(
+        QList(ctx.words.begin() + 1, ctx.words.end()).join('\n'));
+
+    auto message = QStringLiteral("Updated filter rules.");
```
Comment (pajlada) [inline src/controllers/commands/builtin/chatterino/Debugging.cpp:29]: Break the final filter rules out into a local variable & add it to the message

File: src/controllers/commands/builtin/chatterino/Debugging.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    if (ctx.words.size() < 2)
+    {
+        ctx.channel->addMessage(makeSystemMessage(
+            "Usage: /c2:loggingrules <rules...>. To enable debug logging for "
+            "all "
+            "categories from chatterino, use 'chatterino.*.debug=true'. For "
+            "the format on the rules, see "
+            "https://doc.qt.io/qt-6/"
+            "qloggingcategory.html#configuring-categories"));
+        return {};
```
Comment (pajlada) [inline src/controllers/commands/builtin/chatterino/Debugging.cpp:23]: We return `""` at the end of the function - I'm fine with either but we should use the same in both places

File: src/controllers/commands/CommandController.cpp [hunk 4]
```diff
@@ -3211,6 +3212,8 @@ void CommandController::initialize(Settings &, Paths &paths)
 
     this->registerCommand("/shield", &commands::shieldModeOn);
     this->registerCommand("/shieldoff", &commands::shieldModeOff);
+
+    this->registerCommand("/c2:loggingrules", &commands::setLoggingRules);
```
Comment (pajlada) [inline src/controllers/commands/CommandController.cpp:3216]: Naming is the hardest part - I think I would prefer `/c2-set-logging-rules` but happy to discuss it further before merging this in

**Review by pajlada — CHANGES_REQUESTED** [review]
Small nitpicks, otherwise looks good :+1: Thank you

File: CHANGELOG.md [hunk 5]
```diff
@@ -15,6 +15,7 @@
 - Bugfix: Fixed Ctrl+Backspace not working after Select All in chat search popup. (#4461)
 - Bugfix: Fixed crash when scrolling up really fast. (#4621)
 - Dev: Added the ability to control the `followRedirect` mode for requests. (#4594)
+- Dev: Added command to set Qt's logging filter/rules at runtime (`/c2:loggingrules`). (#4637)
```
Comment (Wissididom) [inline CHANGELOG.md:18]: I think that should be in the `Unversioned` section, because 2.4.4 already is released.

Comment (Nerixyz) [reply CHANGELOG.md:18]: Thank you, totally overlooked that.

File: src/controllers/commands/CommandController.cpp (hunk 4 above)
Comment (Nerixyz) [reply src/controllers/commands/CommandController.cpp]: Sounds like a good name to me.

File: src/controllers/commands/builtin/chatterino/Debugging.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+    auto filterRules = QList(ctx.words.begin() + 1, ctx.words.end()).join('\n');
+#else
+    auto filterRules = [&]() {
+        auto tmpList = ctx.words;
+        tmpList.pop_front();
+        return tmpList.join('\n');
+    }();
+#endif
```
Comment (pajlada) [inline src/controllers/commands/builtin/chatterino/Debugging.cpp:33]: Would this be a portable alternative? [qt5](https://doc.qt.io/qt-5/qlist.html#mid) [qt6](https://doc.qt.io/qt-6/qlist.html#mid)

```suggestion
    auto filterRules = ctx.words.mid(1).join('\n');
```

**Review by pajlada — CHANGES_REQUESTED** [review]

**Review by pajlada — APPROVED** [review]

**Comment by pajlada** [comment]
Thank you!

**Pull request merged**